File: app/src/main/java/com/wp/csmu/classschedule/activity/ScoreActivity.java
package com.wp.csmu.classschedule.activity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import com.google.android.material.snackbar.Snackbar;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.appcompat.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.wp.csmu.classschedule.R;
import com.wp.csmu.classschedule.network.NetWorkHelper;
import com.wp.csmu.classschedule.view.adapter.ScoreRecyclerAdapter;
import com.wp.csmu.classschedule.view.bean.Score;
import com.wp.csmu.classschedule.view.utils.BindView;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class ScoreActivity extends BaseActivity implements ScoreRecyclerAdapter.OnClickListener {
    @BindView(R.id.scoreRecyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.scoreToolbar)
    Toolbar toolbar;
    @BindView(R.id.scoreCoordinatorLayout)
    CoordinatorLayout coordinatorLayout;
    @BindView(R.id.scoreSwipeRefreshLayout)
    SwipeRefreshLayout swipeRefreshLayout;
    Handler handler;
    ScoreRecyclerAdapter adapter;
    List<Score> data;

    String[]terms;
    String[]termId;
    String currentTermName;
    String currentTermId;
    int clickedItem;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_score);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        handler = new Handler(new Handler.Callback() {
            @Override
            public boolean handleMessage(Message msg) {
                switch (msg.what) {
                    case 0:
                        swipeRefreshLayout.setRefreshing(false);
                        adapter.updateData(data);
                        recyclerView.scrollToPosition(0);
                        getSupportActionBar().setSubtitle(terms[clickedItem]);
                        if(data.size()==0) {
                            Snackbar.make(coordinatorLayout, "暂无信息", Snackbar.LENGTH_SHORT).show();
                        }
                        break;
                    case -1:
                        swipeRefreshLayout.setRefreshing(false);
                        Exception e = (Exception) msg.obj;
                        Snackbar.make(coordinatorLayout, e.toString(), Snackbar.LENGTH_SHORT).show();
                        break;
                }
                return true;
            }
        });

        swipeRefreshLayout.setColorSchemeResources(R.color.colorPrimary, R.color.colorAccent, R.color.colorPrimaryDark);
        //设置布局管理器
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        //设置动画
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        //初始化适配器
        adapter = new ScoreRecyclerAdapter(new ArrayList<Score>(), this);
        recyclerView.setAdapter(adapter);
        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                getScore1();
            }
        });
        getScore();
    }

    @Override
    public void onClick(View view, int position) {
        View view1 = LayoutInflater.from(this).inflate(R.layout.score_info_dialog_layout, null);
        TextView credit = view1.findViewById(R.id.scoreInfoDialogCredit);
        TextView term=view1.findViewById(R.id.scoreInfoDialogTerm);
        TextView subjectAttr=view1.findViewById(R.id.scoreInfoDialogSubjectAttr);
        TextView examAttr=view1.findViewById(R.id.scoreInfoDialogExamAttr);
        TextView subjectNature=view1.findViewById(R.id.scoreInfoDialogSubjectNature);
        TextView note=view1.findViewById(R.id.scoreInfoDialogNote);
        Score score=data.get(position);
        credit.setText(score.getCredit()+"分");
        term.setText(score.getTerm());
        subjectAttr.setText(score.getSubjectAttribute());
        examAttr.setText(score.getExamAttribute());
        subjectNature.setText(score.getSubjectNature());
        note.setText(score.getNote());
        AlertDialog.Builder dialog = new AlertDialog.Builder(this).setView(view1)
                .setTitle(data.get(position).getSubject()).setNegativeButton("确定", new DialogInterface.OnClickListener(){
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                });
        dialog.show();
    }

    void getScore() {
        SharedPreferences sharedPreferences = getSharedPreferences("user", MODE_PRIVATE);
        final String account = sharedPreferences.getString("account", "");
        swipeRefreshLayout.setRefreshing(true);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    JSONArray jsonArray=NetWorkHelper.getTerms(account);
                    terms=new String[jsonArray.length()];
                    termId=new String[jsonArray.length()];
                    for(int i=0;i<jsonArray.length();i++){
                        JSONObject jsonObject=jsonArray.getJSONObject(i);
                        if ("1".equals(jsonObject.getString("isdqxq"))){
                            clickedItem=i;
                            currentTermName=jsonObject.getString("xqmc");
                            currentTermId=jsonObject.getString("xnxq01id");
                        }
                        terms[i]=jsonObject.getString("xqmc");
                        termId[i]=jsonObject.getString("xnxq01id");
                    }
                    data = NetWorkHelper.getScore(account,currentTermId);
                    handler.sendEmptyMessage(0);
                } catch (Exception e) {
                    e.printStackTrace();
                    Message msg = new Message();
                    msg.what = -1;
                    msg.obj = e;
                    handler.sendMessage(msg);
                }
            }
        }).start();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.score_menu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.scoreMenuSelectTerm:
                selectTerm();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    void selectTerm(){
        final AlertDialog.Builder build=new AlertDialog.Builder(this).setTitle("选择学期").setSingleChoiceItems(terms, clickedItem, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                clickedItem=which;
                getScore1();
                dialog.dismiss();
            }
        }).setNegativeButton("取消", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

            }
        });
        build.show();
    }

    void getScore1() {
        SharedPreferences sharedPreferences = getSharedPreferences("user", MODE_PRIVATE);
        final String account = sharedPreferences.getString("account", "");
        swipeRefreshLayout.setRefreshing(true);
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    data = NetWorkHelper.getScore(account,termId[clickedItem]);
                    handler.sendEmptyMessage(0);
                } catch (Exception e) {
                    e.printStackTrace();
                    Message msg = new Message();
                    msg.what = -1;
                    msg.obj = e;
                    handler.sendMessage(msg);
                }
            }
        }).start();
    }
}

File: app/src/main/java/com/wp/csmu/classschedule/activity/ScoreActivity.kt
package com.wp.csmu.classschedule.activity

import android.app.AlertDialog
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import com.wp.csmu.classschedule.R
import com.wp.csmu.classschedule.network.service.ServiceClient
import com.wp.csmu.classschedule.view.adapter.ScoreRecyclerAdapter
import com.wp.csmu.classschedule.view.bean.Score
import com.wp.csmu.classschedule.view.bean.ScoreFilterBean
import com.wp.csmu.classschedule.view.utils.BindView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.*
import kotlin.collections.LinkedHashSet

class ScoreActivity : BaseActivity(), ScoreRecyclerAdapter.OnClickListener {
    @BindView(R.id.scoreRecyclerView)
    internal var recyclerView: RecyclerView? = null

    @BindView(R.id.scoreToolbar)
    internal var toolbar: Toolbar? = null

    @BindView(R.id.scoreCoordinatorLayout)
    internal var coordinatorLayout: CoordinatorLayout? = null

    @BindView(R.id.scoreSwipeRefreshLayout)
    internal var swipeRefreshLayout: SwipeRefreshLayout? = null
    lateinit var adapter: ScoreRecyclerAdapter
    lateinit var data: LinkedHashSet<Score>

    lateinit var filterBean: ScoreFilterBean
    private val filterTemp = Array(3) { 0 } // 存储用户的筛选选项，学期id、课程性质，显示方式

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_score)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        swipeRefreshLayout!!.setColorSchemeResources(R.color.colorPrimary, R.color.colorAccent, R.color.colorPrimaryDark)
        //设置布局管理器
        recyclerView!!.layoutManager = LinearLayoutManager(this)
        //设置动画
        recyclerView!!.itemAnimator = DefaultItemAnimator()
        //初始化适配器
        adapter = ScoreRecyclerAdapter(LinkedHashSet(), this)
        recyclerView!!.adapter = adapter
        swipeRefreshLayout!!.setOnRefreshListener {
            GlobalScope.launch(Dispatchers.Main) {
                getScore1()
            }
        }
        getScore()
    }

    override fun onClick(view: View, position: Int) {
        val view1 = LayoutInflater.from(this).inflate(R.layout.score_info_dialog_layout, null)
        val credit = view1.findViewById<TextView>(R.id.scoreInfoDialogCredit)
        val term = view1.findViewById<TextView>(R.id.scoreInfoDialogTerm)
        val subjectAttr = view1.findViewById<TextView>(R.id.scoreInfoDialogSubjectAttr)
        val examAttr = view1.findViewById<TextView>(R.id.scoreInfoDialogExamAttr)
        val subjectNature = view1.findViewById<TextView>(R.id.scoreInfoDialogSubjectNature)
        val skillGrade = view1.findViewById<TextView>(R.id.scoreInfoDialogSkillGrade)
        val performanceGrade = view1.findViewById<TextView>(R.id.scoreInfoDialogPerformanceGrade)
        val knowledgePoints = view1.findViewById<TextView>(R.id.scoreInfoDialogKnowledgePoints)
        val (term1, courseName, _, skillScores, performanceScore, knowledgePoints1, credits, examAttribute, subjectNature1, subjectAttribute) = data.elementAt(position)

        credit.text = credits.toString() + "分"
        term.text = term1
        subjectAttr.text = subjectAttribute
        examAttr.text = examAttribute
        subjectNature.text = subjectNature1
        skillGrade.text = skillScores
        performanceGrade.text = performanceScore
        knowledgePoints.text = knowledgePoints1

        if (skillGrade.text.toString() != "" && skillGrade.text.toString().toDouble() < 60.0) skillGrade.setTextColor(Color.RED)
        if (performanceGrade.text.toString() != "" && performanceGrade.text.toString().toDouble() < 60.0) performanceGrade.setTextColor(Color.RED)
        if (knowledgePoints.text.toString() != "" && knowledgePoints.text.toString().toDouble() < 60.0) knowledgePoints.setTextColor(Color.RED)

        val dialog = AlertDialog.Builder(this).setView(view1)
                .setTitle(courseName).setNegativeButton("确定") { _, _ -> }
        dialog.show()
    }

    // [useDefault]=true表示第一次启动时
    private fun getScore() {
        GlobalScope.launch(Dispatchers.Main) {
            swipeRefreshLayout!!.isRefreshing = true
            try {
                filterBean = withContext(Dispatchers.IO) { ServiceClient.getGradeQueryFilter() }

                filterTemp[0] = filterBean.termId.findValueIndex(true)
                filterTemp[1] = filterBean.courseXZ.findValueIndex(true)
                filterTemp[2] = filterBean.displayMode.findValueIndex(true)

                getScore1()
            } catch (e: Exception) {
                e.printStackTrace()
                onOperationFailed(e)
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.score_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.scoreMenuFilter -> {
                if (this::filterBean.isInitialized) {
                    showFilter()
                }
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun showFilter() {
        val filterView = layoutInflater.inflate(R.layout.score_filter_dialog_layout, null)

        filterView.findViewById<Spinner>(R.id.termFilterSpinner).apply { initSpinner(0, this, filterBean.termId) }
        filterView.findViewById<Spinner>(R.id.courseXZFilterSpinner).apply { initSpinner(1, this, filterBean.courseXZ) }
        filterView.findViewById<Spinner>(R.id.displayModeFilterSpinner).apply { initSpinner(2, this, filterBean.displayMode) }

        val dialog = AlertDialog.Builder(this).also {
            it.setView(filterView)
            it.setTitle("筛选")
            it.setPositiveButton("确定") { _, _ ->
                GlobalScope.launch(Dispatchers.Main) {
                    getScore1()
                }
            }
        }
        dialog.setOnCancelListener {
            filterTemp[0] = filterBean.termId.findValueIndex(true)
            filterTemp[1] = filterBean.courseXZ.findValueIndex(true)
            filterTemp[2] = filterBean.displayMode.findValueIndex(true)
        }
        dialog.show()
    }

    private fun initSpinner(index: Int, spinner: Spinner, item: LinkedHashMap<Pair<String, String>, Boolean>) {
        val adapter = ArrayAdapter(
                this@ScoreActivity,
                android.R.layout.simple_spinner_item,
                item.keys.map { it.second }
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter
        spinner.setSelection(filterTemp[index])
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                filterTemp[index] = position
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {

            }
        }
    }

    private suspend fun getScore1(): Boolean {
        if (!this::filterBean.isInitialized) {
            getScore()
            return true
        }
        try {
            withContext(Dispatchers.Main) { swipeRefreshLayout!!.isRefreshing = true }
            data = withContext(Dispatchers.IO) {
                Log.i("ScoreActivity", "paused")
                if (filterBean.courseXZ.keys.elementAt(filterTemp[1]).first == "") { // 如果查询全部性质成绩
                    val list = filterBean.courseXZ.keys.map { it.first }.drop(1) as ArrayList
                    ServiceClient.queryAllXZGrades(
                            filterBean.termId.keys.elementAt(filterTemp[0]).first,
                            list,
                            filterBean.displayMode.keys.elementAt(filterTemp[2]).first
                    )
                } else {
                    ServiceClient.queryGrades(
                            filterBean.termId.keys.elementAt(filterTemp[0]).first,
                            filterBean.courseXZ.keys.elementAt(filterTemp[1]).first,
                            filterBean.displayMode.keys.elementAt(filterTemp[2]).first
                    )
                }
            }
            withContext(Dispatchers.Main) { onOperationSucceed() }
            return true
        } catch (e: Exception) {
            e.printStackTrace()
            withContext(Dispatchers.Main) {
                onOperationFailed(e)
            }
            return false
        }
    }

    private fun onOperationSucceed() {
        swipeRefreshLayout!!.isRefreshing = false
        adapter.updateData(data)
        recyclerView!!.scrollToPosition(0)
        filterBean.termId.assignElementTrue(filterTemp[0])
        filterBean.courseXZ.assignElementTrue(filterTemp[1])
        filterBean.displayMode.assignElementTrue(filterTemp[2])
        setToolbarSubtitle()
        if (data.size == 0) {
            Snackbar.make(coordinatorLayout!!, "暂无信息", Snackbar.LENGTH_SHORT).show()
        }
    }

    private fun setToolbarSubtitle() {
        val term = filterBean.termId.findKey(true)?.second
        val courseXZ = filterBean.courseXZ.findKey(true)?.second
        val displayMode = filterBean.displayMode.findKey(true)?.second
        supportActionBar!!.subtitle = "$term / $courseXZ / $displayMode"
    }

    private fun onOperationFailed(reason: Exception) {
        swipeRefreshLayout!!.isRefreshing = false
        Snackbar.make(coordinatorLayout!!, reason.toString(), Snackbar.LENGTH_SHORT).show()
    }

    // 寻找map中第一次出现value的index
    private fun <K, V> Map<K, V>.findValueIndex(value: V): Int {
        var index = 0
        this.entries.forEachIndexed { i, entry ->
            if (entry.value == value) {
                index = i
                return@forEachIndexed
            }
        }
        return index
    }

    private fun <K, V> Map<K, V>.findKey(value: V): K? {
        var key: K? = null
        this.entries.forEach { entry ->
            if (entry.value == value) {
                key = entry.key
                return@forEach
            }
        }
        return key
    }

    private fun <K> HashMap<K, Boolean>.assignElementTrue(index: Int) {
        var count = 0
        this.forEach {
            this[it.key] = count == index
            count++
        }
    }
}
